overlord/devicestate/handlers.go

```diff
@@ -698,3 +703,231 @@ func fetchKeys(st *state.State, keyID string) (errAcctKey error, err error) {
 		keyID = a.SignKeyID()
 	}
 }
+
+func snapState(st *state.State, name string) (*snapstate.SnapState, error) {
+	var snapst snapstate.SnapState
+	err := snapstate.Get(st, name, &snapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return &snapst, nil
+}
+
+func makeRollbackDir(name string) (string, error) {
+	snapRollbackDir := filepath.Join(dirs.SnapRollbackDir, name)
+
+	if err := os.MkdirAll(dirs.SnapRollbackDir, 0755); err != nil {
+		return "", err
+	}
+
+	if err := os.Mkdir(snapRollbackDir, 0755); err != nil {
+		// fail even if directory exists already
+		return "", err
+	}
+
+	return snapRollbackDir, nil
+}
+
+func currentGadgetInfo(snapst *snapstate.SnapState) (*gadget.Info, error) {
```

Comment:
Perhaps release.OnClassic should be a formal argument so that the caller decides and there is no extra mocking necessary to correctly test this function?

Reply:
that's the approach ReadGadgetInfo itself follows

Reply:
sorry, I missed this when I commented on this issue, are we going to support gadget updates at all on classic atm ?  we support gadget on classics but for things like slots or hooks, not boot assets.  If we don't then this code should just hardcode classic false actually,  and the task itself error super early if for some mysterious reason is been used on classic

Reply:
Hardcoding false and erroring out early sounds reasonable.

Reply:
Pushed a patch to address this.